### PR TITLE
feat: add materials for BTV tutorial & add array output options

### PR DESCRIPTION
**Pull request by Ming-Yan**
**New features**
- runner: add new options to store arrays, not output hist, overwrite
- utils: add systematics
- utils: add roccor, switch jerc, add jet mask files
- workflow: add array output and systematic axis
- helper: add array output
- scripts: add script for hist coffea->TH1/2D root

**Tutorial materials** 
- add test coffea file to demonstrate plotter

**Fixes**
- plot : minor fix on plotting scripts
- ci: split ttbar DL/SL & remove ttcom&validation workflow
- ci: remove custom corrections check
- ci: make ci work with ssh
- remove redundunt files
- scripts: rename directory
- wf: minor fixes for DY
- utils: minor fixes
- doc: add one section for developers

**code from external source** 
-  [PDF/scale weights](https://github.com/cms-btv-pog/BTVNanoCommissioning/blob/569b46151030421bb77a01139c8b4774cc9f411a/src/BTVNanoCommissioning/utils/correction.py#L1180-L1300) are taken from https://github.com/andrzejnovak/boostedhiggs/blob/master/boostedhiggs/corrections.py#L58-L145 
-  [filefetcher.py](https://github.com/cms-rwth/CoffeaRunner/pull/18) updated with more customized way from -> no modification needed for the dictionary keys

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,7 +74,10 @@ More options for `runner.py`
                         files{ "Rereco17_94X","Winter22Run3","2018_UL","2017_UL","2016preVFP_UL","2016postVFP_UL"}
   --isCorr              Run with SFs
   --isJERC              JER/JEC implemented to jet
-  --isSyst              Run with systematics for SF
+  --isSyst              Run with systematics, all, weights_only(no JERC uncertainties included),JERC_split, None(not extract )
+  --isArray             Output root files
+  --noHist              Not save histogram coffea files
```
[the remark this answers is not in the document]
Comment (Ming-Yan) [reply README.md:79]: then if we specify this we only store the cutflow sumw only.

File: runner.py [hunk 2]
```diff
@@ -696,12 +741,28 @@ def get_main_parser():
...
+    if args.isArray:
+        if args.overwrite and path.exists(
+            args.output.replace(".coffea", "").replace("hists_", "")
+        ):
+            os.system(
+                f'rm -r {args.output.replace(".coffea", "").replace("hists_", "")}'
+            )
+        os.mkdir(args.output.replace(".coffea", "").replace("hists_", ""))
+        os.system(
+            f'mv tmp/*.root {args.output.replace(".coffea", "").replace("hists_","")}/.'
```
[the remark this answers is not in the document]
Comment (Ming-Yan) [reply runner.py:763]: They are just the branches of the stored variables

File: scripts/fetch.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        if dataset.startswith("#") or dataset.strip() == "":
+            # print("we skip this line:", line)
+            continue
+
+        dsname = dataset.strip().split("/")[1]  # Dataset first name
+
+        Tier = dataset.strip().split("/")[
+            3
+        ]  # NANOAODSIM for regular samples, USER for private
+        if "SIM" not in Tier:
```
[the remark this answers is not in the document]
Comment (Ming-Yan) [reply scripts/fetch.py:59]: it's the same as `NANOAODSIM`

File: src/BTVNanoCommissioning/utils/correction.py [hunk 4]
```diff
@@ -301,173 +345,955 @@ def add_jec_variables(jets, event_rho):
...
+            jet.btagDeepFlavCvB = ak.fill_none(jet.btagDeepFlavCvB, 0.0)
+            jet.btagDeepCvL = ak.fill_none(jet.btagDeepCvL, 0.0)
+            jet.btagDeepCvB = ak.fill_none(jet.btagDeepCvB, 0.0)
+            jet.hadronFlavour = ak.fill_none(jet.hadronFlavour, 0)
+            if "correctionlib" in str(type(correct_map["ctag"])):
+                if SFtype == "DeepJetC":
+                    tmp_sfs = np.where(
+                        masknone,
+                        1.0,
+                        correct_map["ctag"]["deepJet_shape"].evaluate(
```
[the remark this answers is not in the document]
Comment (Ming-Yan) [reply src/BTVNanoCommissioning/utils/correction.py:650]: no, this is not yet supported

File: src/BTVNanoCommissioning/utils/xs_scaler.py [hunk 5]
```diff
@@ -57,7 +59,9 @@ def scaleSumW(output, lumi):
                                 / scaled[files][sample]["sumw"]
                             )
                         else:
-                            if not (("data" in sample) or ("Run" in sample)):
+                            if not (("data" in sample) or ("Run" in sample)) or (
+                                "Double" in sample
```
[the remark this answers is not in the document]
Comment (Ming-Yan) [reply src/BTVNanoCommissioning/utils/xs_scaler.py:63]: indeed this is temporary modification for my analysis code, remove it.

File: src/BTVNanoCommissioning/workflows/example.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
+import collections, awkward as ak, numpy as np
```
[the remark this answers is not in the document]
Comment (Ming-Yan) [reply src/BTVNanoCommissioning/workflows/example.py:1]: Yeah, I think it's a good idea. I can put in the notes for developer.

File: src/BTVNanoCommissioning/workflows/__init__.py [hunk 7]
```diff
@@ -43,6 +43,10 @@
     NanoProcessor as CTAGSemilepttValidSFProcessor,
 )
 
+# from BTVNanoCommissioning.workflows.example import (
+#     NanoProcessor as ExampleProcessor,
+# )
+
 # FIXME - make names more systematic?
 workflows = {}
 workflows["validation"] = ValidationProcessor
```
[the remark this answers is not in the document]
Comment (demuller) [reply src/BTVNanoCommissioning/workflows/__init__.py:52]: True, I agree, maybe we can rename it to `release_validation.py` to make it clearer what the use-case of this workflow is?

File: src/BTVNanoCommissioning/workflows/ctag_DY_valid_sf.py [hunk 8]
```diff
@@ -178,174 +215,193 @@ def process(self, events):
...
-                        )
-                        jetsfs_b[0][f"SF_{syst}_dn"] = btagSFs(
-                            sjets[:, 0], self.SF_map, "DeepJetB", f"down"
-                        )
+        # Systematics information
+        if shift_name is None:
+            systematics = ["nominal"] + list(weights.variations)
+        else:
+            systematics = [shift_name]
+        exclude_btv = [
```
[the remark this answers is not in the document]
Comment (demuller) [reply src/BTVNanoCommissioning/workflows/ctag_DY_valid_sf.py:251]: Ah, yes, sorry I have misread these variables as b-tagging discrimants. thx for clarification!

File: src/BTVNanoCommissioning/utils/correction.py [hunk 9]
```diff
@@ -301,173 +345,956 @@ def add_jec_variables(jets, event_rho):
...
+    top = gen[(gen.pdgId == 6) & gen.hasFlags(["isLastCopy"])]
+    anti_top = gen[(gen.pdgId == -6) & gen.hasFlags(["isLastCopy"])]
+    return np.sqrt(
+        top_pT_sf_formula(ak.flatten(top.pt, axis=-1))
+        * top_pT_sf_formula(ak.flatten(anti_top.pt, axis=-1))
+    )
+
+
+# Jennet adds PS weights
+# https://github.com/andrzejnovak/boostedhiggs/blob/master/boostedhiggs/corrections.py#L88-L108
```
Comment (demuller) [inline src/BTVNanoCommissioning/utils/correction.py:1232]: please add also to the description of this PR, which parts of correction.py have been taken from the external source

**Review by demuller — COMMENTED** [review]
Hi @Ming-Yan I have only one final comment, otherwise this PR looks very nice!

File: src/BTVNanoCommissioning/utils/correction.py (hunk 9 above)
Comment (Ming-Yan) [reply src/BTVNanoCommissioning/utils/correction.py:1232]: Thanks for review the PR, I updated the comments now :)

**Review by demuller — APPROVED** [review]

**Pull request merged**